File: ppp_prediction/model/glmnet.py
import pandas as pd
from ppp_prediction.metrics import cal_binary_metrics
import numpy as np


# current
def run_glmnet(
    train,
    xvar,
    label,
    time=None,
    test=None,
    covariate=None,
    cv=5,
    alpha=1,
    lambda_=None,
    trace_it=1,
    family="gaussian",
    weights=None,
    type_measure="deviance",
    coef_choice="lambda.min",
    standardize=True,
    intercept=False,
    save_path=None,
):
    import rpy2.robjects as robjects

    from rpy2.robjects import pandas2ri
    from rpy2.robjects.packages import importr
    from rpy2.robjects.conversion import localconverter
    import pandas as pd

    glmnet_script_path = __file__.replace("glmnet.py", "glmnet.R")
    robjects.r.source(glmnet_script_path)
    glmnet_func = robjects.r["glmnet_lasso"]
    pandas2ri.activate()
    # format conversion
    if covariate is not None and isinstance(covariate, list):
        covariate = robjects.StrVector(covariate)
    if xvar is not None and isinstance(xvar, list):
        xvar = robjects.StrVector(xvar)
    if weights is not None:
        if isinstance(weights, str):
            if weights == "balanced":
                from sklearn.utils import class_weight

                weights = class_weight.compute_sample_weight(
                    class_weight="balanced", y=train[label]
                )

        if isinstance(weights, pd.Series):
            weights = robjects.FloatVector(weights.values)
        elif isinstance(weights, np.ndarray):
            weights = robjects.FloatVector(weights)
        else:
            weights = robjects.FloatVector(weights)

    with localconverter(robjects.default_converter + pandas2ri.converter):
        result = glmnet_func(
            train,
            xvar,
            label,
            time if time is not None else robjects.r("NULL"),
            test if test is not None else robjects.r("NULL"),
            covariate if covariate is not None else robjects.r("NULL"),
            cv,
            alpha,
            lambda_ if lambda_ is not None else robjects.r("NULL"),
            trace_it,
            family,
            type_measure,
            coef_choice,
            standardize,
            intercept,
            weights=robjects.r("NULL") if weights is None else weights,
            save_path=save_path if save_path is not None else robjects.r("NULL"),
        )
        # result = [list(i) for i in result]
        train_mean = result["train_mean"]
        train_std = result["train_std"]
        train = result["train"]
        if train is not None:
            train = train.rename(columns={"pred": f"{label}_pred"})
            # train_metrics =
            try:
                to_cal_train = train[[label, f"{label}_pred"]].dropna()
                train_metrics = cal_binary_metrics(
                    to_cal_train[label], to_cal_train[f"{label}_pred"], ci=True
                )
            except:
                train_metrics = None
        if test is not None:
            # test.rename(columns={"pred": f"{label}_pred"})
            test = result["test"].rename(columns={"pred": f"{label}_pred"})
            try:
                to_cal_test = test[[label, f"{label}_pred"]].dropna()
                test_metrics = cal_binary_metrics(
                    to_cal_test[label], to_cal_test[f"{label}_pred"], ci=True
                )
            except:
                test_metrics = None
        else:
            test = None
        coef = result["coef"]
        # rm intercept
        keeped = coef.index.str.contains(".*Intercept.*")
        coef = coef[~keeped]
    model = glmnet_linear(
        coef,
        feature_names=xvar,
        mean=train_mean,
        std=train_std,
    )
    # train_metrics = cal_binary_metrics(train["y"], train["y_hat"])

    return (
        model,
        train_metrics,
        test_metrics,
        train,
        test,
    )


class glmnet_linear:
    def __init__(self, coef, intercept=None, feature_names=None, mean=None, std=None):
        self.coef = coef
        self.intercept = intercept if intercept is not None else 0
        self.feature_names = feature_names
        self.mean = mean
        self.std = std

    def predict(self, X):
        if isinstance(X, pd.DataFrame):
            if self.feature_names is not None:
                X = X[self.feature_names]
            else:
                X = X.values
        elif isinstance(X, pd.Series):
            X = X.values
        if self.mean is not None and self.std is not None:
            X = (X - self.mean) / self.std

        return X @ self.coef + self.intercept

File: ppp_prediction/model/glmnet.R
options(repos = c(CRAN = "https://cloud.r-project.org/"))

# 检查并安装必要的包
packages <- c("glmnet", "survival", "ggplot2", "export", "svglite", 'rjson', 'optparse', 'arrow', "dplyr")

for (pkg in packages) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    if (pkg == "BiocManager") {
      install.packages("BiocManager")
    }
    BiocManager::install(pkg)
  }
}

suppressPackageStartupMessages({
  library(glmnet)
  library(survival)
  library(ggplot2)
  library(export)
  library(svglite)
  library(rjson)
  library(arrow)
  library(dplyr)
library(optparse)

})
sumweights <- function(data, coef, xvar) {
    # 确保 coef 是一个向量, xvar 是用来指定数据和系数中的列
    coef_vector <- as.numeric(coef[xvar,])
    
    # 应用函数计算每行的加权和
    return(apply(data[, xvar], 1, function(x) sum(x * coef_vector)))
}

standardize_func <- function(data, xvar = NULL, means = NULL, sds = NULL) {
  # 计算每列的均值和标准差
  if (is.null(xvar)) xvar <- colnames(data)
  if (is.null(means)) means <- colMeans(data[, xvar])
  if (is.null(sds)) sds <- apply(data[, xvar], 2, sd)

  means <- means[xvar]
  sds <- sds[xvar]

  data[, xvar] <- data.frame(scale(data[, xvar], center = means, scale = sds))
  # 标准化数据
  return (list(
    data =data,
    mean = means,
    std = sds
  ))

}

glmnet_lasso<-function(
  train,
  xvar,
  label,
  time=NULL,
  test = NULL,
  covariate = NULL,
  cv = 5,
  alpha=1,
  weights = NULL,
  lambda = NULL,
  trace.it = 1,
  family = "gaussian",
  type.measure = "deviance",
  coef_choice = "lambda.min",
  standardize  = TRUE,
  intercept = FALSE,
  parallel=TRUE,
  save_path = NULL
){
  # drop na
  set.seed(2077)
  used_fatures = xvar
  p.fac <- rep(1, length(xvar))
  print("begin")

  if (!is.null(covariate)) {
    used_fatures = c(used_fatures, covariate)
    p.fac <- c(p.fac, rep(0, length(covariate)))
  }
  print(length(p.fac) == length(used_fatures))

  if (family == "cox"){
    if (is.null(time)){
      stop("time is required for cox model")
    }
    train = train[complete.cases(train[, c(used_fatures, label, time)]), ]
  } else{
    train = train[complete.cases(train[, c(used_fatures, label)]), ]
  }

  if (standardize){
    standardize <- standardize_func(train, used_fatures)
    train_mean = standardize$mean
    train_std = standardize$std
    train = standardize$data

    if (!is.null(test)) {
      test = standardize_func(test, used_fatures)$data
    }


  }else{
    train_mean = NULL
    train_std = NULL
  }

  print(sprintf("train data size: %d with featuers %d", nrow(train), length(used_fatures)))

  if (family == "cox") {
    train_y = train[c(label, time)]
    colnames(train_y) = c("status", "time")

    cvfit.cv = cv.glmnet(
      as.matrix(train[, used_fatures]),
      as.matrix(train_y),
      alpha = alpha,
      weights=weights,
      nfolds = cv,
      lambda = lambda,
      trace.it = trace.it,
      family = "cox",
      type.measure = "C",
      penalty.factor = p.fac,
      parallel = parallel,
      standardize = F,
      intercept = intercept
    )
  }else{
  cvfit.cv = cv.glmnet(
    as.matrix(train[, used_fatures]),
    as.matrix(train[, label]),
    alpha = alpha,
    nfolds = cv,
    lambda = lambda,
    trace.it = trace.it,
    family = family,
    type.measure = type.measure,
    penalty.factor = p.fac,
    parallel = parallel,
    standardize = F,
    intercept = intercept
  )
  }


  coef_ = coef(cvfit.cv, s = coef_choice)
  train$pred = sumweights(train, coef_, xvar)
  if (!is.null(test)) {
    test$pred = sumweights(test, coef_, xvar)
  }
  if (!is.null(save_path)) {
    saveRDS(cvfit.cv, file = save_path)
  }
  return (
    list(
      # cvfit = cvfit.cv,
      coef = as.data.frame(as.matrix(coef_)),
      train = train,
      test = test,
      train_mean = as.data.frame(as.matrix(train_mean)),
      train_std = as.data.frame(as.matrix(train_std))
    )
  )
  # print("end")
  # return (
  #   list(
  #     cvfit = cvfit.cv,
  #     coef = coef_,
  #     train = train,
  #     test = test,
  #     train_mean = train_mean,
  #     train_std = train_std
  #   )
  # )
}